Src11 security information standard (#55)
FuelLabs/sway-standards: standards/src11-security-information/README.md

        <source media="(prefers-color-scheme: dark)" srcset=".docs/src-11-logo-dark-theme.png">
        <img alt="SRC-11 logo" width="400px" src=".docs/src-11-logo-light-theme.png">
    </picture>
</p>

# Abstract

The following standard allows for contract creators to make communication information readily available to everyone, with the primary purpose of allowing white hat hackers to coordinate a bug-fix or securing of funds.

# Motivation

White hat hackers may find bugs or exploits in contracts that they want to report to the project for safeguarding of funds. It is not immediately obvious from a ContractId, who the right person to contact is. This standard aims to make the process of bug reporting as smooth as possible

# Prior Art

The [security.txt](https://github.com/neodyme-labs/solana-security-txt) library for Solana has explored this idea. This standard takes inspiration from the library, with some changes.

# Specification

## Security Information Type

The following describes the `SecurityInformation` type. 

- The struct MAY contain `None` for `Option<T>` type fields, if they are deemed unnecessary.
- The struct MUST NOT contain empty `String` or `Vec` fields.

> - The struct MAY contain a link or the information directly for the following fields: `project_url`, `policy`, `encryption`, `source_code`, `auditors`, `acknowledgments`, `additional_information`.

Reviewer: ```suggestion
- The struct MAY contain a URL or the information directly for the following fields: `project_url`, `policy`, `encryption`, `source_code`, `auditors`, `acknowledgments`, `additional_information`.
```

> - The struct MUST contain the information directly for the following fields: `name`, `contact_information`, `preferred_languages`, `source_release`, and `source_revision`.

Reviewer: Not sure how I feel about "directly". Maybe "The struct MUST contain the following information encoded as a `String` for the following fields:"?

<p align="center">
    <picture>
        <source media="(prefers-color-scheme: dark)" srcset=".docs/src-6-logo-dark-theme.png">
        <img alt="SRC-6 logo" width="400px" src=".docs/src-6-logo-light-theme.png">
    </picture>
</p>

# Abstract

The following standard allows for contract creators to make communication information readily available to everyone, with the primary purpose of allowing white hat hackers to coordinate a bug-fix or securing of funds.

# Motivation

> White hat hackers may find bugs or exploits in contracts which they want to report to the project for safeguarding of funds. It is not immedietely obvious from a ContractId, who the right person to contact is. This standard aims to make the process of bug reporting as smooth as possible

Reviewer: ```suggestion
White hat hackers may find bugs or exploits in contracts that they want to report to the project for safeguarding of funds. It is not immediately obvious from a ContractId, who the right person to contact is. This standard aims to make the process of bug reporting as smooth as possible
```

# Prior Art

> The [security.txt](https://github.com/neodyme-labs/solana-security-txt) library for solana has explored this idea. This standard takes inspiration from the library, with some changes.

Reviewer: ```suggestion
The [security.txt](https://github.com/neodyme-labs/solana-security-txt) library for Solana has explored this idea. This standard takes inspiration from the library, with some changes.
```

# Specification

## Required public functions

The following function MUST be implemented to follow the SRC-11 standard.

### `fn security_information() -> SecurityInformation;`

> This function takes no input parameters and returns a struct containing contact information for the project owners, information regarding the bug bounty program, other information related to security, and any other information which the developers find relevant.

Reviewer: ```suggestion
This function takes no input parameters and returns a struct containing contact information for the project owners, information regarding the bug bounty program, other information related to security, and any other information that the developers find relevant.
```

- This function MUST return accurate and up to date information.

> - This function MAY not return the optional parameters.

Reviewer: ```suggestion
- This function MAY return the optional parameters.
```
Wouldn't this be part of `SecurityInformation`

> - This function MUST return atleast one item in the `contact_information` field's `Vec`. Furthurmore, the string should follow the following format `<contact_type>:<contact_information>`.

Reviewer: ```suggestion
- This function MUST return at least one item in the `contact_information` field's `Vec`. Furthermore, the string should follow the following format `<contact_type>:<contact_information>`.
```

> - This function MUST NOT return any empty strings. Optional parameters must return `None` if it is unnecessary.

Reviewer: ```suggestion
- This function MUST NOT return any empty strings. Optional parameters MUST return `None` if it is unnecessary.
```

> - This function MUST return atleast one item in the `preferred_languages`, if it is not `None`. Furthurmore, the string should only contain the `ISO 639-1` language code and nothing else.

Reviewer: ```suggestion
- This function MUST return at least one item in the `preferred_languages`, if it is not `None`. Furthermore, the string should only contain the `ISO 639-1` language code and nothing else.
```

Perhaps a link to ISO 639-1's definition would be relevant here

> - This function MAY return a link or the information directly for the following fields: `project_url`, `policy`, `encryption`, `source_code`, `auditors`, `acknowledgements`, `additional_information`.

Reviewer: ```suggestion
- This function MAY return a link or the information directly for the following fields: `project_url`, `policy`, `encryption`, `source_code`, `auditors`, `acknowledgments`, `additional_information`.
```

> - This function MUST return the information directly for the following fields: `name`, `contact_information`, `preferred_languages`, `source_release`, `source_revision`.

Reviewer: ```suggestion
- This function MUST return the information directly for the following fields: `name`, `contact_information`, `preferred_languages`, `source_release`, and `source_revision`.
```

# Rationale

The return structure discussed covers most information that may want to be conveyed regarding the security of the contract, with an additional field to convey any additional information. This should allow easy communication between the project owners and any white hat hackers if necessary.

# Backwards Compatibility

This standard does not face any issues with backward compatibility. 

> // Is this section necessary?

Reviewer: Is this meant to be here?

# Security Considerations

> The information is entirely self reported and as such might not be accurate. Accuracy of information cannot be enforced and as such, anyone using this information should be aware of that.

Reviewer: > "might not be accurate"

Isn't it defined in the standard that it must be up to date?

Author: Yeah but people can lie lol, important to note that its entirely self reported, just because its on the blockchain doesnt make it verified or real
